### PR TITLE
Support retrying functions when exceptions occur.

### PR DESCRIPTION
**Pull request by rtamalin**
Add retry_on_exception() which can retry an arbitrary function call, encapsulated as a functools.partial, when an exception occurs.

Wrap all our hook calls with retry_on_exception().

Fix typo in get_now() log call.

Add unit tests for retry_on_exception().

Fixes: #81

### CODE REVIEW

File: csp_billing_adapter/utils.py [hunk 1]
```diff
@@ -133,3 +136,127 @@ def get_prev_bill_time(
...
+    """
+
+    if logger is None:
+        # if no specific logger was provided use default log
+        logger = log
+
+    if func_name is None:
+        # if no function name is provided try using the name of
+        # func_call.func, or failing that, 'unnamed_function'
+        func_name = getattr(
```
Comment (smarlowucf) [inline csp_billing_adapter/utils.py:191]: Why wouldn't a function name be provided?

**Review by smarlowucf — COMMENTED** [review]
Looks good but I think there's a couple functions that don't benefit from retries. Anything that doesn't potentially rely on I/O where intermittent issues would be expected.

File: csp_billing_adapter/utils.py (hunk 1 above)
Comment (rtamalin) [reply csp_billing_adapter/utils.py:191]: So for most functions, the dynamic lookup of func_call.func.__name__ works, but for hook calls this attribute doesn't exist, so I plumbed in being able to provide a func_name explicitly for use with the hook calls, but for most other cases the dynamic retrieval will work.

Comment (rtamalin) [reply csp_billing_adapter/utils.py:191]: We potentially could leverage retry_on_exception() in any of the plugins by importing it from csp_billing_adapter.utils, and in such cases, when not dealing with hook calls, passing a func_name wouldn't be required anymore.

Comment (rtamalin) [reply csp_billing_adapter/utils.py:191]: And we could implement a decorator based upon retry_on_exception() if that would be beneficial...

File: csp_billing_adapter/config.py [hunk 2]
```diff
@@ -83,7 +87,14 @@ def load_defaults(data, hook):
         """
 
         defaults = {}
-        hook.load_defaults(defaults=defaults)
```
[the remark this answers is not in the document]
Comment (rtamalin) [reply csp_billing_adapter/config.py:86]: Yeah no problem; happy to remove.

I figured I would start by wrapping all the hook calls, and then for any that we were sure we didn't need to wrap, the wrapping can be easily removed.

[the remark this answers is not in the document]
Comment (rtamalin) [reply csp_billing_adapter/config.py]: Removed

File: csp_billing_adapter/utils.py (hunk 1 above)
Comment (smarlowucf) [reply csp_billing_adapter/utils.py:191]: Seems good for now. :+1:

File: csp_billing_adapter/config.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (smarlowucf) [reply csp_billing_adapter/config.py]: Yeah, doesn't hurt to have it wrapped either. Can add it back if necessary.

**Review by smarlowucf — APPROVED** [review]

**Pull request merged**